All macros require the output to implement Display

Because `try_scan!` has the following line
```rs
format_args!($pattern, $($arg),*);
```
an implementation of `Display` is mandatory for any type that you want to convert to.

Example:
```rs
enum Foo {
    Bar,
    Baz,
}

impl FromStr for Foo {
    type Err = ();

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "Bar" => Ok(Foo::Bar),
            "Baz" => Ok(Foo::Baz),
            _ => Err(()),
        }
    }
}

fn main() {
    let foo: Foo = read!();
}
```
Something like the former would be expected to work, but because of the issue, `Foo` would be required to implement `Display` as well. This is unnecessary in most cases, so removing that line would make this library easier to use.

Here is the error the compiler gives;
```rs
error[E0277]: `Foo` doesn't implement `std::fmt::Display`
   --> src\main.rs:106:20
    |
106 |     let foo: Foo = read!();
    |                    ^^^^^^^ `Foo` cannot be formatted with the default formatter
    |
    = help: the trait `std::fmt::Display` is not implemented for `Foo`
    = note: in format strings you may be able to use `{:?}` (or {:#?} for pretty-print) instead
    = note: this error originates in the macro `format_args` which comes from the expansion of the macro `read` (in Nightly builds, run with -Z macro-backtrace for more info)
```

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -174,9 +174,11 @@ macro_rules! try_scan(
     ($input:expr => $pattern:expr, $($arg:expr),*) => {{
         $crate::try_scan!(@impl question_mark; $input => $pattern, $($arg),*)
     }};
+    // implementation detail.
     (@question_mark: $($e:tt)+) => {{
         ($($e)+)?
     }};
+    // implementation detail.
     (@unwrap: $($e:tt)+) => {{
         ($($e)+).unwrap()
     }};
@@ -207,7 +209,7 @@ macro_rules! try_scan(
             $crate::try_scan!(@$action: match_next(c, stdin))
         }
 
-        format_args!($pattern, $($arg),*);
+        format_args!($pattern, $({let _ = &$arg; ""}),*);
     }};
 );
 
@@ -232,3 +234,24 @@ macro_rules! scan(
         $crate::try_scan!(@impl unwrap; $input => $pattern, $($arg),*)
     }};
 );
+
+#[cfg(test)]
+mod tests {
+    use std::convert::Infallible;
+
+    use super::*;
+
+    /// dummy struct that does not implement Display
+    #[derive(Debug)]
+    struct NoDisplay;
+    impl FromStr for NoDisplay {
+        type Err = Infallible;
+        fn from_str(_: &str) -> Result<Self, Self::Err> {
+            Ok(Self)
+        }
+    }
+    #[test]
+    fn test_no_display() {
+        let _: NoDisplay = read!("{}", "".bytes());
+    }
+}

